api/src/main/java/org/openmrs/module/reporting/dataset/definition/evaluator/IterableSqlDataSetEvaluator.java

```diff
@@ -0,0 +1,135 @@
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.openmrs.annotation.Handler;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.reporting.common.ObjectUtil;
+import org.openmrs.module.reporting.common.SqlRunner;
+import org.openmrs.module.reporting.dataset.DataSet;
+import org.openmrs.module.reporting.dataset.IterableSqlDataSet;
+import org.openmrs.module.reporting.dataset.definition.DataSetDefinition;
+import org.openmrs.module.reporting.dataset.definition.IterableSqlDataSetDefinition;
+import org.openmrs.module.reporting.evaluation.EvaluationContext;
+import org.openmrs.module.reporting.evaluation.EvaluationException;
+import org.openmrs.module.reporting.common.SqlIterator;
+import org.openmrs.module.reporting.evaluation.querybuilder.SqlQueryBuilder;
+import org.openmrs.util.OpenmrsUtil;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.util.Iterator;
+import java.util.Properties;
+
+/**
+ * The logic that evaluates a {@link IterableSqlDataSetDefinition} and produces an {@link DataSet}
+ *
+ * @see IterableSqlDataSetDefinition
+ */
+@Handler(supports = {IterableSqlDataSetDefinition.class})
+public class IterableSqlDataSetEvaluator implements IterableDataSetEvaluator {
+
+    private static final Logger log = LoggerFactory.getLogger(IterableSqlDataSetEvaluator.class);
+
+    /**
+     * Public constructor
+     */
+    public IterableSqlDataSetEvaluator() {
+    }
+
+    /**
+     * @should evaluate a IterableSqlDataSetDefinition
+     * @should evaluate a IterableSqlDataSetDefinition with parameters
+     * @should evaluate a IterableSqlDataSetDefinition with in statement
+     * @should protect SQL Query Against database modifications
+     * @see DataSetEvaluator#evaluate(DataSetDefinition, EvaluationContext)
+     */
+    public DataSet evaluate(DataSetDefinition dataSetDefinition, EvaluationContext context) throws EvaluationException {
+
+        context = ObjectUtil.nvl(context, new EvaluationContext());
+
+        IterableSqlDataSetDefinition sqlDsd = (IterableSqlDataSetDefinition) dataSetDefinition;
+
+        SqlQueryBuilder queryBuilder = new SqlQueryBuilder();
+
+        String sqlQuery = sqlDsd.getSql();
+
+        queryBuilder.append(sqlQuery);
+
+        queryBuilder.setParameters(context.getParameterValues());
+
+        IterableSqlDataSetDefinition defenition = (IterableSqlDataSetDefinition) dataSetDefinition;
+        Properties connectionProperties = getConnectionProperties(defenition.getConnectionPropertyFile());
+        Iterator iterator = null;
+        Connection connection = null;
+        try {
+            connection = createConnection(connectionProperties);
+            SqlRunner runner = new SqlRunner(connection);
+
+            if (StringUtils.isNotBlank(defenition.getSql())) {
```

Comment:
As far as I can tell, the definition supports "sql", "sqlFile", and "sqlResource" as means to get the sql to evaluate.  But here you seem only to be checking the "sql".  Don't you need some higher-level utility method to pull the SQL based on which of these properties is populated?  What does the SqlFileDataSetEvaluator do?

Reply:
You can see how they are implemented in the SqlFileDataSetEvaluator class.  They are straightforward - they support either a String property, a file on the class path (sqlResource), or a file on the file system (sqlFile).

Reply:
See my comment above.  We should implement these - they are straightforward to implement, and if you end up extending the SqlFileDataSetDefinition (which I think you should), then you'll need to do so.  Otherwise, if you don't intend to implement them, then don't add them as configuration properties to your definition class.  Why add them as properties if you don't intend to support them?

Reply:
Done.